### PR TITLE
 Make FFT size and oversampling adjustable in smbPitchShifter, add windowing to smbFFT

### PR DESCRIPTION
**Pull request by dahlgrenmartin**

### CODE REVIEW

File: servers/audio/effects/audio_effect_spectrum_analyzer.cpp [hunk 1]
```diff
@@ -81,9 +81,10 @@ void AudioEffectSpectrumAnalyzerInstance::process(const AudioFrame *p_src_frames
 
 		float *fftw = temporal_fft.ptrw();
 		for (int i = 0; i < to_fill; i++) { //left and right buffers
-			fftw[(i + temporal_fft_pos) * 2] = p_src_frames[i].l;
+			float window = -.5 * Math::cos(2. * Math_PI * (double)i / (double)to_fill) + .5;
```
Comment (akien-mga) [inline servers/audio/effects/audio_effect_spectrum_analyzer.cpp]: Could you use fully qualified floats, i.e. `0.5` and `2.0`? That makes code more readable IMO and is the usual way we write them in this codebase.

File: servers/audio/effects/audio_effect_pitch_shift.cpp [hunk 2]
```diff
@@ -293,8 +293,10 @@ void AudioEffectPitchShiftInstance::process(const AudioFrame *p_src_frames, Audi
 	float *out_l = (float *)p_dst_frames;
 	float *out_r = out_l + 1;
 
-	shift_l.PitchShift(base->pitch_scale, p_frame_count, 2048, 4, sample_rate, in_l, out_l, 2);
-	shift_r.PitchShift(base->pitch_scale, p_frame_count, 2048, 4, sample_rate, in_r, out_r, 2);
+	static const int fft_sizes[AudioEffectPitchShift::FFT_Size::FFT_SIZE_MAX] = { 256, 512, 1024, 2048, 4096 };
```
Comment (akien-mga) [inline servers/audio/effects/audio_effect_pitch_shift.cpp]: GCC 5 seems to fail on this:
```
servers/audio/effects/audio_effect_pitch_shift.cpp: In member function 'virtual void AudioEffectPitchShiftInstance::process(const AudioFrame*, AudioFrame*, int)':
servers/audio/effects/audio_effect_pitch_shift.cpp:296:52: error: 'AudioEffectPitchShift::FFT_Size' is not a class or namespace
  static const int fft_sizes[AudioEffectPitchShift::FFT_Size::FFT_SIZE_MAX] = { 
                                                    ^
```
I think you should be able to use `FFT_SIZE_MAX` directly here.

File: servers/audio/effects/audio_effect_pitch_shift.cpp [hunk 3]
```diff
@@ -315,14 +317,51 @@ float AudioEffectPitchShift::get_pitch_scale() const {
...
+void AudioEffectPitchShift::set_fft_size(FFT_Size p_fft_size) {
+	ERR_FAIL_INDEX(p_fft_size, FFT_SIZE_MAX);
+	fft_size = p_fft_size;
+}
+
+AudioEffectPitchShift::FFT_Size AudioEffectPitchShift::get_fft_size() const {
+	return fft_size;
+}
+
+
```
Comment (akien-mga) [inline servers/audio/effects/audio_effect_pitch_shift.cpp]: Style issue here, `clang-format` doesn't allow more than 1 consecutive empty lines.

File: servers/audio/effects/audio_effect_pitch_shift.h [hunk 4]
```diff
@@ -81,15 +81,27 @@ class AudioEffectPitchShiftInstance : public AudioEffectInstance {
 
 public:
 	virtual void process(const AudioFrame *p_src_frames, AudioFrame *p_dst_frames, int p_frame_count);
+
```
Comment (akien-mga) [inline servers/audio/effects/audio_effect_pitch_shift.h]: Style issue here too as per `clang-format`'s style.

**Comment by akien-mga** [comment]
See [Code style guidelines](http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html) for details on the style guide and how to apply `clang-format` locally.

And [PR workflow](http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase) on how to rebase your branch to squash the current two commits together and do necessary fixups with `git commit --amend`.

**Comment by akien-mga** [comment]
For the reference, this is a re-do of #28547.

**Comment by mdahlgrengadd** [comment]
@akien-mga After about one million screwups with git rebase... Is this ok?

**Comment by akien-mga** [comment]
It looks good yes, thanks :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**